Systems Audit post (#80)
RockefellerArchiveCenter/bits_and_bytes: _posts/2020-03-25-lessons-learned-systems-audit.md

---
title: "Lessons Learned from a Systems Audit"
date: 2020-03-25T14:00:00+00:00
author: Patrick Galligan
layout: post
categories:
  - Software and Systems
tags:
  - systems
  - continuous integration
  - continuous deployment
  - maintenance

excerpt_separator: <!--more-->
---

> We’ve been doing _a lot_ of technological systems work at the Archive Center over the past few years; we’ve been standing up and migrating to new systems, tweaking and refining existing systems, and developing entirely new services to help with everyday work. We’ve recently kicked off a systems infrastructure audit project that aims to help us understand our systems better and make improvements on how we develop for, deploy, and maintain our archival systems. This blog posts aims to provide some background for why we took a close look at all of our systems, both in production and development, and lays out some findings from our systems audit, and some of the general talking points that we posed to ourselves in trying to figure out a way forward. Hopefully, our experiences will help others in the same situation figure out what questions they should be asking of themselves.

Reviewer: ```suggestion
We’ve been doing _a lot_ of technological systems work at the Archive Center over the past few years; we’ve been standing up and migrating to new systems, tweaking and refining existing systems, and developing entirely new services to help with everyday work. We’ve recently kicked off a systems infrastructure audit project that aims to help us understand our systems better and make improvements on how we develop, deploy, and maintain our archival systems. This blog posts aims to provide some background for why we took a close look at all of our systems, both in production and development, and lays out some findings from our systems audit, and some of the general talking points that we posed to ourselves in trying to figure out a way forward. Hopefully, our experiences will help others in the same situation figure out what questions they should be asking of themselves.
```

Reviewer: (removed `for` after `developed`)

![RAC systems over time]({{site.baseurl}}/assets/img/2020/03/systems-chart.png)

Coming out of the data gathering, I tried to synthesize what I was seeing about our systems into general talking points for a future team meeting about the results of the survey. I broke the talking points into three general sections of questions and talking points: Systems in Production, Server Infrastructure, and Systems Maintenance. Below are some of the most salient questions and talking points.

  *	Systems in Production
    *	How can we generally improve our management of our systems?

>     *	Are there opportunities to have IT help with our systems load? Vice versa?

Reviewer: ```suggestion
    *	Are there opportunities to collaborate more effectively with IT?
```
Trying to be political 😉

  *	Server Updates and Distribution
    *	How can we better stay on top of updating our production and development servers?
    *	Whose responsibility should it be to keep on top of distribution and package updates and management?

>     *	What is our deployment pattern for Continuous Deployment?

Reviewer: ```suggestion
    *	Is Continuous Deployment a useful pattern for us?
```

Reviewer: This concept (Continuous Deployment) could use a link as I suspect it will be new to many readers.

  *	Systems Maintenance
    *	What are our current systems maintenance practices? Do they exist in a codified form?
    *	Where are the current systems/application management breakdowns?

The D-Team had a lively discussion that ran for almost two hours just trying to unpack what we were currently doing and what we saw as the next steps forward. Perhaps unsurprisingly to those with systems experience, we identified dependency management and updates as one of our most time-consuming, manual, and frustrating operations. We have so many different applications that use a multitude of different packages that just making sure all of the packages stay up to date is a time-consuming task right now. Additionally, we’re somewhat caught between two worlds: our newer, internally-developed systems have good test coverage and we’ve been moving towards Docker containers for deployment, but our procedures for older systems and applications often fall short.

> We think we could lessen some of these pains by moving iteratively towards Continuous Integration (CI) and Continuous Deployment (CD) processes, but we need to get a better understanding of what we’re already doing and how that would fit into future pipelines first. In order to move towards more comprehensive CI/CD practices, we’ll also need better automated test coverage on all of our systems, systems resource tracking, some form of logging analysis, and SSH key management. We could also go back and start containerizing all of our older applications so that they will be easier to redeploy when necessary. We’re in the early stages of exploring tools to help us take our pipelines to the next stage, but we’re open to suggestions, so if you have experience in doing this at a cultural heritage institution, we’d love to hear about it!

Reviewer: This is definitely an opportunity for some links
